// @progress/kendo-vue-layout/dist/npm/menu/components/MenuItemArrow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuItemArrowVue2 = exports.MenuItemArrow = exports.leftArrowIcon = exports.rightArrowIcon = exports.downArrowIcon = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var misc_1 = require("../utils/misc");
/**
 * @hidden
 */
exports.downArrowIcon = kendo_svg_icons_1.caretAltDownIcon;
/**
 * @hidden
 */
exports.rightArrowIcon = kendo_svg_icons_1.caretAltRightIcon;
/**
 * @hidden
 */
exports.leftArrowIcon = kendo_svg_icons_1.caretAltLeftIcon;
/**
 * @hidden
 */
var MenuItemArrowVue2 = {
  name: 'KendoMenuItemArrow',
  props: {
    itemId: String,
    dir: String,
    verticalMenu: Boolean
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    return h("span", {
      "class": "k-menu-expand-arrow",
      "aria-hidden": "true",
      attrs: this.v3 ? undefined : {
        "aria-hidden": "true"
      }
    }, [h(kendo_vue_common_1.Icon, {
      icon: this.getArrowName(),
      attrs: this.v3 ? undefined : {
        icon: this.getArrowName(),
        role: "presentation"
      },
      role: "presentation"
    })]);
  },
  methods: {
    getArrowName: function getArrowName() {
      var childrenPosition = (0, misc_1.getChildrenPosition)(this.$props.itemId, this.$props.verticalMenu === true, this.$props.dir === 'rtl');
      return childrenPosition === 'downward' ? exports.downArrowIcon : childrenPosition === 'rightward' ? exports.rightArrowIcon : childrenPosition === 'leftward' ? exports.leftArrowIcon : '';
    }
  }
};
exports.MenuItemArrowVue2 = MenuItemArrowVue2;
/**
 * @hidden
 */
var MenuItemArrow = MenuItemArrowVue2;
exports.MenuItemArrow = MenuItemArrow;
