// @progress/kendo-vue-dropdowns/dist/esm/common/ListContainer.js
var __assign = this && this.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Popup } from '@progress/kendo-vue-popup';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var ListContainerVue2 = {
  name: 'list-container',
  // @ts-ignore
  emits: {
    mousedown: null,
    blur: null,
    open: null,
    close: null
  },
  props: {
    width: [String, Number],
    dir: String,
    itemsCount: Number,
    popupSettings: {
      type: Object,
      default: function _default() {
        return {
          animate: true,
          height: '200px'
        };
      }
    }
  },
  created: function created() {
    this.kendoAnchorRef = undefined;
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      this.$emit('mousedown', e);
    },
    onBlur: function onBlur(e) {
      this.$emit('blur', e);
    },
    onOpen: function onOpen(e) {
      this.$emit('open', e);
    },
    onClose: function onClose(e) {
      this.$emit('close', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    var _a = this.$props,
      onMouseDown = _a.onMouseDown,
      onBlur = _a.onBlur,
      width = _a.width,
      dir = _a.dir,
      popupSettings = _a.popupSettings;
    // @ts-ignore
    var popupClass = popupSettings.popupClass,
      className = popupSettings.className,
      animate = popupSettings.animate,
      anchor = popupSettings.anchor,
      anchorAlign = popupSettings.anchorAlign,
      show = popupSettings.show,
      open = popupSettings.open,
      close = popupSettings.close,
      itemsCount = popupSettings.itemsCount,
      appendTo = popupSettings.appendTo,
      collision = popupSettings.collision,
      direction = popupSettings.direction,
      offset = popupSettings.offset,
      popupAlign = popupSettings.popupAlign,
      style = popupSettings.style,
      popupWidth = popupSettings.width,
      height = popupSettings.height,
      popupStyle = popupSettings.popupStyle;
    return (
      // @ts-ignore function children
      h(Popup, {
        ref: 'popup',
        style: __assign(__assign({}, style), {
          width: popupWidth ? popupWidth : width,
          direction: dir,
          height: height
        }),
        popupClass: className,
        attrs: this.v3 ? undefined : {
          popupClass: className,
          animate: animate,
          anchor: anchor,
          anchorAlign: anchorAlign,
          show: show,
          contentKey: itemsCount,
          appendTo: appendTo,
          collision: collision,
          direction: direction,
          offset: offset,
          popupAlign: popupAlign
        },
        animate: animate,
        anchor: anchor,
        anchorAlign: anchorAlign,
        show: show,
        contentKey: itemsCount,
        onOpen: this.onOpen,
        on: this.v3 ? undefined : {
          "open": this.onOpen,
          "close": this.onClose
        },
        onClose: this.onClose,
        appendTo: appendTo,
        collision: collision,
        direction: direction,
        offset: offset,
        popupAlign: popupAlign
      }, this.v3 ? function () {
        return [h("div", {
          "class": popupClass,
          style: popupStyle,
          onMousedown: _this.onMouseDown,
          on: _this.v3 ? undefined : {
            "mousedown": _this.onMouseDown,
            "focusout": _this.onBlur
          },
          onFocusout: _this.onBlur
        }, [defaultSlot])];
      } : [h("div", {
        "class": popupClass,
        style: popupStyle,
        onMousedown: _this.onMouseDown,
        on: _this.v3 ? undefined : {
          "mousedown": _this.onMouseDown,
          "focusout": _this.onBlur
        },
        onFocusout: _this.onBlur
      }, [defaultSlot])])
    );
  }
};
/**
 * @hidden
 */
var ListContainer = ListContainerVue2;
export { ListContainer, ListContainerVue2 };
