// @progress/kendo-vue-grid/dist/esm/header/GridHeaderSelectionCell.js
var __assign = this && this.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { guid, getTemplate } from '@progress/kendo-vue-common';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { messages, selectAllRows } from '../messages/main.js';
/**
 * @hidden
 */
var GridHeaderSelectionCellVue2 = {
  props: {
    field: String,
    title: String,
    selectionValue: Boolean,
    render: [Object, Function, String]
  },
  created: function created() {
    this.inputId = guid();
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  methods: {
    changeHandle: function changeHandle(e) {
      this.$emit('selectionchange', {
        field: this.$props.field,
        event: e
      });
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var renderTemplate = this.$props.render;
    var ls = provideLocalizationService(this);
    var selectAllRowsMessage = ls.toLanguageString(selectAllRows, messages[selectAllRows]);
    var input = h("input", {
      checked: this.v3 ? this.$props.selectionValue : null,
      domProps: this.v3 ? undefined : {
        "checked": this.$props.selectionValue
      },
      id: this.inputId,
      attrs: this.v3 ? undefined : {
        id: this.inputId,
        type: "checkbox",
        "aria-label": selectAllRowsMessage
      },
      type: "checkbox",
      "class": "k-checkbox k-checkbox-md k-rounded-md",
      "aria-label": selectAllRowsMessage,
      onChange: this.changeHandle,
      on: this.v3 ? undefined : {
        "change": this.changeHandle
      }
    });
    var defaultRendering = getTemplate.call(this, {
      h: h,
      template: renderTemplate,
      defaultRendering: input,
      additionalProps: __assign(__assign({}, this.$props), {
        selectAll: selectAllRowsMessage
      }),
      additionalListeners: {
        selectionchange: this.changeHandle
      }
    });
    return h("span", {
      "class": "k-link"
    }, [h("span", {
      "class": "k-column-title"
    }, [h("span", {
      "class": "k-checkbox-wrap"
    }, [defaultRendering])])]);
  }
};
/**
 * @hidden
 */
var GridHeaderSelectionCell = GridHeaderSelectionCellVue2;
export { GridHeaderSelectionCell, GridHeaderSelectionCellVue2 };
